compiler/rustc_codegen_ssa/src/base.rs

```diff
@@ -124,6 +124,23 @@ fn unsized_info<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
             let old_info =
                 old_info.expect("unsized_info: missing old info for trait upcasting coercion");
             if data_a.principal_def_id() == data_b.principal_def_id() {
+                // Codegen takes advantage of the additional assumption, where if the
+                // principal trait def id of what's being casted doesn't change,
+                // then we don't need to adjust the vtable at all. This
+                // corresponds to the fact that `dyn Tr<A>: Unsize<dyn Tr<B>>`
+                // requires that `A = B`; we don't allow *upcasting* objects
+                // between the same trait with different args. If we, for
+                // some reason, were to relax the `Unsize` trait, it could become
+                // unsound, so let's assert here that the trait refs are *equal*.
+                //
+                // We can use `assert_eq` because the binders should have been anonymized,
+                // and because higher-ranked equality now requires the binders are equal.
+                debug_assert_eq!(
+                    data_a.principal(),
+                    data_b.principal(),
```

Comment:
No but there is not an upcast for which this *is* valid. That would literally be unsound in the type system.

Reply:
Like, checking projection validity seems more appropriate as an assertion on the Unsize trait goal in the trait system or something, but even that seems excessive. 

Like, the only reason miri needs to check projections are compatible is because of transmutes.

Reply:
And also because I don't trust the `library/` and would prefer this to be double-checked in the compiler. ;)

But maybe that's overly paranoid, and anyway it's a different PR. I just wanted to understand what is and is not being checked here, and the comments should reflect that.

Reply:
> And also because I don't trust the library/ and would prefer this to be double-checked in the compiler. ;)

It's not really `library/`, it's just the soundness of the `Unsize` goal, and the fact that you can't impl `impl Foo for Bar { type Assoc = A; }` and `impl Foo for Bar { type Assoc = B; }`. That is, normalization is a function.